File: server/src/main/java/cn/iinti/atom/system/SysLogAspect.java
package cn.iinti.atom.system;


import cn.iinti.atom.entity.SysLog;
import cn.iinti.atom.entity.UserInfo;
import cn.iinti.atom.mapper.SysLogMapper;
import cn.iinti.atom.service.base.alert.EventNotifierService;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import proguard.annotation.Keep;

import org.apache.commons.lang3.StringUtils;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.Arrays;


@Slf4j
@Aspect
@Component
public class SysLogAspect {

    @Resource
    private SysLogMapper sysLogMapper;

    @Resource
    private EventNotifierService eventNotifierService;

    @Pointcut("@annotation(cn.iinti.atom.system.LoginRequired)")
    @Keep
    public void pointcut() {
    }

    @Keep
    @Around("pointcut()")
    public Object around(ProceedingJoinPoint point) throws Throwable {
        Object result = point.proceed();
        // 保存日志
        saveLog(point);
        return result;
    }


    private void saveLog(ProceedingJoinPoint point) {
        UserInfo user = AppContext.getUser();
        if (user == null) {
            return;
        }
        if (AppContext.isApiUser()) {
            return;
        }

        LoginRequired loginAnnotation = AppContext.getLoginAnnotation();
        if (loginAnnotation.skipLogRecord()) {
            return;
        }
        MethodSignature signature = (MethodSignature) point.getSignature();
        String className = point.getTarget().getClass().getName();
        String methodName = signature.getName();
        String target = className + "#" + methodName;

        Method method = signature.getMethod();
        SysLog sysLog = new SysLog();
        sysLog.setUsername(AppContext.getUser().getUserName());
        sysLog.setOperation(method.getName());
        String params = Arrays.toString(point.getArgs());
        sysLog.setParams(StringUtils.substring(params, 0, 50));

        sysLog.setMethodName(target);
        log.info("record sys log:{}", sysLog);
        sysLogMapper.insert(sysLog);

        // alert
        if (loginAnnotation.alert()) {
            eventNotifierService.notifySensitiveOperation(sysLog.getUsername(),
                    target, sysLog.getParams()
            );
        }
    }


}

File: server/src/main/kotlin/cn/iinti/katom/system/SysLogAspect.kt
package cn.iinti.katom.system

import cn.iinti.katom.entity.SysLog
import cn.iinti.katom.mapper.SysLogMapper
import cn.iinti.katom.base.alert.EventNotifierService
import jakarta.annotation.Resource
import org.apache.commons.lang3.StringUtils
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.stereotype.Component
import java.lang.reflect.Method
import java.util.*

@Aspect
@Component
class SysLogAspect {
    @Resource
    private lateinit var sysLogMapper: SysLogMapper

    @Resource
    private lateinit var eventNotifierService: EventNotifierService

    // 手动实现Lombok的@Slf4j功能
    private val log = org.slf4j.LoggerFactory.getLogger(SysLogAspect::class.java)

    @Around("@annotation(cn.iinti.katom.system.LoginRequired)")
    @Throws(Throwable::class)
    fun around(point: ProceedingJoinPoint): Any? {
        val result = point.proceed()
        // 保存日志
        saveLog(point)
        return result
    }

    private fun saveLog(point: ProceedingJoinPoint) {
        val user = AppContext.getUser()
        if (user == null || AppContext.isApiUser()) {
            return
        }

        val loginAnnotation = AppContext.getLoginAnnotation()
        if (loginAnnotation?.skipLogRecord == true) {
            return
        }

        val signature = point.signature as MethodSignature
        val className = point.target.javaClass.name
        val methodName = signature.name
        val target = "$className#$methodName"

        val method: Method = signature.method
        val sysLog = SysLog()
        sysLog.username = AppContext.getUser()!!.userName
        sysLog.operation = method.name
        val params = Arrays.toString(point.args)
        sysLog.params = StringUtils.substring(params, 0, 50)

        sysLog.methodName = target
        log.info("record sys log:{}", sysLog)
        sysLogMapper.insert(sysLog)

        // alert
        if (loginAnnotation?.alert == true) {
            eventNotifierService.notifySensitiveOperation(
                sysLog.username!!,
                target,
                sysLog.params!!
            )
        }
    }
}

